### PR TITLE
[Backport release-24.05] traefik: 3.0.4 -> 3.1.2, fix CVE-2024-45410

### PR DESCRIPTION
**Pull request by Scrumplex**
## Description of changes

This PR updates Traefik to the latest version that targeted Go 1.22, as Go 1.23 isn't availabe in 24.05 (yet?). Then https://github.com/Scrumplex/traefik/commit/21f0062a6d570f12ce51646cacdcb28178d8be6f is applied, which is the upstream fix for the vulnerability, but rebased on v3.1.2. As confirmed in https://github.com/traefik/traefik/pull/11077#issuecomment-2371307267 applying this patch should fix the issue.

This fixes CVE-2024-45410[[0]] (GHSA-62c8-mh53-4cqv[[1]]).

Indirect backport of https://github.com/NixOS/nixpkgs/pull/344149

Relevant Releases:

https://github.com/traefik/traefik/releases/tag/v3.1.3

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc
[0]: https://nvd.nist.gov/vuln/detail/CVE-2024-45410
[1]: https://github.com/traefik/traefik/security/advisories/GHSA-62c8-mh53-4cqv

### CODE REVIEW

**Comment by Scrumplex** [comment]
https://github.com/ofborg test traefik

File: pkgs/servers/traefik/default.nix [hunk 1]
```diff
@@ -1,17 +1,26 @@
...
 
-  vendorHash = "sha256-4zI4OL4UlaaefxRPQoUtjyn9M8yfuYOBOyYh6vTnMJg=";
+  patches = [
+    # https://github.com/traefik/traefik/pull/11077
+    (fetchpatch2 {
+      name = "traefik-fix-CVE-2024-45410.patch";
+      url = "https://github.com/Scrumplex/traefik/commit/21f0062a6d570f12ce51646cacdcb28178d8be6f.patch";
+      hash = "sha256-Va0NYYAZcHVIRPm0CSpTUDPkPraaqJa65ttHUJR2ME8=";
+    })
+  ];
```
[the remark this answers is not in the document]
Comment (mweinelt) [reply pkgs/servers/traefik/default.nix]: Again: if you prefer updating, then try reverting the Go 1.23 commit instead.

**Comment by Scrumplex** [comment]
I have create a Flake that tests the vulnerability in Traefik for current nixos-unstable, current nixos-24.05 and this PR's branch. nixos-unstable and this PR succeed, while current nixos-24.05 fails.

Try it out yourself:

https://codeberg.org/Scrumplex/Traefik-CVE-2024-45410-Test

```nix
nix flake check --verbose --keep-going git+https://codeberg.org/Scrumplex/Traefik-CVE-2024-45410-Test
```

I was thinking about contributing this test to Nixpkgs, too, but I think upstream already has tests to ensure this doesn't regress.

**Review by cafkafk — APPROVED** [review]
> I was thinking about contributing this test to Nixpkgs, too, but I think upstream already has tests to ensure this doesn't regress.

It could be useful for a time to ensure patches aren't removed preemptively, but would likely soon just make tests take longer, I lean either way on that.

---

I've verified that your test example does work on `x86_64-linux` as expected, and that the other NixOS traefik tests doesn't fail from this. This looks good to me, and a severity 9.3 should probably be expedited.

**Comment by cafkafk** [comment]
For anyone unsure about the patch, here is the diff between upstream and scrumplex's fix https://0x0.st/XgPE.txt

here are the raw links to the patches 
- upstream: https://github.com/traefik/traefik/commit/584144100524277829f26219baaab29a53b8134f.patch 
- scrumplex (this pr): https://github.com/NixOS/nixpkgs/blob/3002cd2380d272a295efa174b3148cb5b7d632e2/pkgs/servers/traefik/fix-CVE-2024-45410.patch

**Comment by Scrumplex** [comment]
Another way to compare my patch:

```console
$ git clone https://github.com/traefik/traefik.git
$ cd traefik

 # Backport upstream v2 -> v3 merge that contains fix
 # See https://github.com/traefik/traefik/commit/093989fc14f40fb54c0fe1221b363427f6129be7
$ git switch -c traefik-3.1.2-backport-merge v3.1.2
 # --strategy-option=ours resolves a conflict in GH Actions manifests
$ git cherry-pick --strategy-option=ours -m1 093989fc14f40fb54c0fe1221b363427f6129be7

 # Apply my patch from this PR
$ git switch -c traefik-3.1.2-backport-Scrumplex v3.1.2
$ curl -L "https://github.com/Scrumplex/nixpkgs/raw/3002cd2380d272a295efa174b3148cb5b7d632e2/pkgs/servers/traefik/fix-CVE-2024-45410.patch" | git am

 # Diff my patch against branch we prepared before
$ git diff traefik-3.1.2-backport-merge -- \
    docs/content/reference/static-configuration/cli-ref.md \
    docs/content/reference/static-configuration/env-ref.md \
    docs/content/reference/static-configuration/file.toml \
    docs/content/reference/static-configuration/file.yaml \
    docs/content/routing/entrypoints.md \
    integration/fixtures/headers/connection_hop_by_hop_headers.toml \
    integration/headers_test.go \
    pkg/config/static/entrypoints.go \
    pkg/middlewares/auth/connectionheader.go \
    pkg/middlewares/auth/connectionheader_test.go \
    pkg/middlewares/auth/forward.go \
    pkg/middlewares/forwardedheaders/forwarded_header.go \
    pkg/middlewares/forwardedheaders/forwarded_header_test.go \
    pkg/middlewares/headers/headers.go \
    pkg/server/server_entrypoint_tcp.go
```

This should show you the diff of all files affected by my patch as well as upstream's v2 patch.

in my case I can only see some changes related to acme and a change that isn't present in the original patch in `pkg/server/server_entrypoint_tcp.go`

**Comment by cafkafk** [comment]
Comparing v3.1.2 to v3.1.3 also confirms that removals make perfect sense https://github.com/traefik/traefik/compare/v3.1.2...v3.1.3

The bump from v3.0 to v3.1 does require minor migrations with limited scope https://doc.traefik.io/traefik/v3.1/migration/v3/#v30-to-v31, but aside from this being fine to backport since its a security release, major version bumps have in the past also been made without reservation https://github.com/NixOS/nixpkgs/pull/307794.

I think this looks solid.

**Pull request merged**